docs: add HA section to stream aggregation (#11006)
VictoriaMetrics/VictoriaMetrics: docs/victoriametrics/stream-aggregation/README.md

  keep_metric_names: true
```

## Scaling aggregation horizontally

> The aggregation output is only correct when every input sample contributing to it lands on the same aggregator process.

Reviewer: ```suggestion
Aggregation output is only correct when all contributing samples are processed by the same aggregator instance.
```

> To scale the aggregation horizontally always shard the input samples in a deterministic way. This can be achieved by

Reviewer: ```suggestion
To scale the aggregation horizontally, always shard the input samples in a deterministic way. This can be achieved by
```

> buidling two layer topology of vmagents where the first layer is responsible for sharding, and the second layer is responsible for aggregating:

Reviewer: ```suggestion
building a two layer topology of vmagents where the first layer is responsible for sharding, and the second layer is responsible for aggregating:
```

```mermaid
flowchart LR
    V1[vmagent-shard-1] -- requests_total{env=test, pod=foo} --> SV1[vmagent-aggr-1]

>     V1[vmagent-shard-1] -- requests_total{env=test, pod=bar} --> SV2[vmagent-aggr-1]

Reviewer: Shouldn't this sample go to `vmagent-aggr-1`?

If I undestand correctly both env=test samples should go to the same aggregator.

Here's the rendered diagram for reference.

<img width="1634" height="424" alt="Image" src="https://github.com/user-attachments/assets/193e060f-35f6-4814-b816-949ec92d0a49" />

Author: Yes, thanks for spotting!

    V2[vmagent-shard-2] -- requests_total{env=prod, pod=baz} --> SV2[vmagent-aggr-2]
    SV1 -- requests_total:5m_without_pod_total{env=test} --> x(( ))
    SV2 -- requests_total:5m_without_pod_total{env=prod} --> y(( ))
style x fill:none,stroke:none
style y fill:none,stroke:none
```

The sharding layer of vmagents can be configured via the `-remoteWrite.shardByURL.labels` or `-remoteWrite.shardByURL.ignoreLabels`
command line flags. See how to [shard data across remote write destinations](https://docs.victoriametrics.com/victoriametrics/vmagent/#sharding-among-remote-storages) for more details.

> There are the following requirements for sharded aggregation to be correct:

Reviewer: ```suggestion
The following requirements must be met for sharded aggregation to work correctly:
```

> - All sharding vmagents should have the same deterministic sharding configuraton.

Reviewer: ```suggestion
- All sharding vmagents should have the same deterministic sharding configuration.
```

> - The sharding configuration must align with the `by` and `without` lists:

Reviewer 2: from this statement and examples below, some might make a wrong assumption that `by` set should be equal to `-remoteWrite.shardByURL.labels`. probably it's better to explicitly mention that `by` should be a **subset** of `-remoteWrite.shardByURL.labels`.

Author: I don't know how to formulate it in a short form. Do you have ideas?

Author: Updated in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11006/commits/b61b2253ca5cfacc9efe05d16ab1cc8b666c6388

> - Sharding configuration should be accounted for in the `by` and `without` lists:

Reviewer: ```suggestion
- The sharding configuration must align with the `by` and `without` lists:
```

  - If you aggregate `by: env` - make sure sharding agents are configured with `-remoteWrite.shardByURL.labels=env`. 
    This makes sure that all the samples for the same `env` are aggregated together and produce the complete output.
  - If you aggregate `without: pod` - make sure sharding agents are configured with `-remoteWrite.shardByURL.ignoreLabels=pod`.

>     This makes sure that `requests_total{env=test, pod=foo}` and `requests_total{env=test, pod=bar}` will be routed to the same aggregator

Reviewer: ```suggestion
    This makes sure that `requests_total{env=test, pod=foo}` and `requests_total{env=test, pod=bar}` are routed to the same aggregator
```

>     and will be aggregated together. See also [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5938#issuecomment-2018470324).

Reviewer: ```suggestion
    and are aggregated together. See also [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/5938#issuecomment-2018470324).
```

- Aggregating vmagents should not produce collisions: the aggregation output should be unique across all the sharded agents.
  For example, `requests_total:5m_without_env_pod_total` produced by both `vmagent-aggr-1` and `vmagent-aggr-2` will collide

>   unless they have labels uniquley identifying them. These labels should be either preserved during sharding and aggregation config,

Reviewer: ```suggestion
  unless they have labels uniquely identifying them. These labels should be either preserved during sharding and aggregation config,
```

  or enforced on the output via `-remoteWrite.label` - see [these docs](https://docs.victoriametrics.com/victoriametrics/stream-aggregation/#cluster-mode) for more details.

> > Never shard histograms by `le` label. A histogram is a logical group of series differing

Reviewer 2: ```suggestion
> Never shard histograms by `le` (or `vmrange` in case of VM histograms) labels. A histogram is a logical group of series differing
```

only in the bucket label. All of those buckets must land on the same aggregator at the same time so it can produce a
coherent bucket set. See more about [aggregating histograms](https://docs.victoriametrics.com/stream-aggregation/#aggregating-histograms).

> See also [why you shouldn't put aggregator behind load balancer](https://docs.victoriametrics.com/stream-aggregation/#put-aggregator-behind-load-balancer).

Reviewer: ```suggestion
See also [why you shouldn't put an aggregator behind a load balancer](https://docs.victoriametrics.com/stream-aggregation/#put-aggregator-behind-load-balancer).
```

The following requirements must be met for sharded aggregation to work correctly:
- All sharding vmagents should have the same deterministic sharding configuration.
- The sharding configuration must align with the `by` and `without` lists:

>   - If you aggregate `by: env` - make sure that `env` label is listed in the routing key of sharding agents: `-remoteWrite.shardByURL.labels=env`. 

Reviewer 2: - Aggregator's `by` labels should be a subset of shard's `-remoteWrite.shardByURL.labels`: with `-remoteWrite.shardByURL.labels=env,job` aggregator's `by` should only include `env`, `job` or both.
- Aggregators's `without` labels should be a superset of shard's `-remoteWrite.shardByURL.ignoreLabels`: with `-remoteWrite.shardByURL.ignoreLabels=env,job` aggregator's `without` should include at least both `env` and `job`

Reviewer 2: don't know how main statement can be improved to be short and clear at the same time, but examples could be changed to something as i wrote above

Author: Updated in https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11006/commits/b61b2253ca5cfacc9efe05d16ab1cc8b666c6388
